torch/distributed/algorithms/ddp_comm_hooks/__init__.py

```diff
@@ -1,23 +1,31 @@
+from enum import Enum
+from functools import partial
+
 import torch.distributed.algorithms.ddp_comm_hooks.default_hooks as default
 import torch.distributed.algorithms.ddp_comm_hooks.quantization_hooks as quantization
 
-# hook_registry wraps the hooks of ``torch.distributed.algorithms.ddp_comm_hooks``
-# library inside a dictionary. After importing hook_registry user can simply register
-# a hook to a ``ddp_model``` using ``hook_registry(comm_hook)(ddp_model, process_group)``.
-hook_registry = {
-    "quantize per tensor": lambda model, pg: model._register_comm_hook(
-        pg, quantization.quantization_pertensor_hook
-    ),
-    "quantize per channel": lambda model, pg: model._register_comm_hook(
-        pg, quantization.quantization_perchannel_hook
-    ),
-    "allreduce": lambda model, pg: model._register_comm_hook(
-        pg, default.allreduce_hook
-    ),
-    "allgather then aggregate": lambda model, pg: model._register_comm_hook(
-        pg, default.allgather_then_aggregate_hook
-    ),
-    "fp16 compress": lambda model, pg: model._register_comm_hook(
-        pg, default.fp16_compress_hook
-    ),
-}
+# Just a wrapper for DDPCommHookType Enum class to store hooks as partial functions.
+def ddp_comm_hook_wrapper(comm_hook, model, state):
+    model._register_comm_hook(state, comm_hook)
+
+# DDPCommHookType enumerates the hooks of ``torch.distributed.algorithms.ddp_comm_hooks``
+# as names and ``ddp_comm_hook_wrapper`` partials with hook specified. As an example, you can
+# register allreduce hook by ``DDPCommHookType.ALLREDUCE.value(model=model, state=process_group)``.
+class DDPCommHookType(Enum):
+    ALLREDUCE = partial(ddp_comm_hook_wrapper, comm_hook=default.allreduce_hook)
+    FP16_COMPRESS = partial(ddp_comm_hook_wrapper, comm_hook=default.fp16_compress_hook)
+    QUANTIZE_PER_TENSOR = partial(
+        ddp_comm_hook_wrapper, comm_hook=quantization.quantization_pertensor_hook
+    )
+    QUANTIZE_PER_CHANNEL = partial(
+        ddp_comm_hook_wrapper, comm_hook=quantization.quantization_perchannel_hook
+    )
+
+# Registers the hooks of ``torch.distributed.algorithms.ddp_comm_hooks`` by taking ``comm_hook_name``
+# input as a ``str```. Also, checks whether hook is in ``DDPCommHookType``.
+def register_ddp_comm_hook(comm_hook_name, model, state=None):
```

Comment:
This API should take as input a DDPCommHookType and not the name. The user should invoke this API as follows: `register_ddp_comm_hook(DDPCommHookType.FP16_COMPRESS, model)`. This makes the API less buggy where the user might misspell names as strings.

Also, add a proper python docstring to this method instead of just the "#" style comments above.

Reply:
Can we still keep a wrapper to be used for the benchmarks?